File: EqSat/src/assembler/iced_amd64_assembler.rs
use crate::assembler::amd64_assembler::IAmd64Assembler;
use iced_x86::code_asm::*;
use iced_x86::Instruction;
use iced_x86::Register;

/// x86-64 assembler implementation using the iced-x86 library
pub struct IcedAmd64Assembler {
    assembler: CodeAssembler,
}

impl IcedAmd64Assembler {
    pub fn new() -> Result<Self, IcedError> {
        Ok(Self {
            assembler: CodeAssembler::new(64)?,
        })
    }

    fn conv(reg: Register) -> AsmRegister64 {
        match reg {
            Register::RAX => rax,
            Register::RCX => rcx,
            Register::RDX => rdx,
            Register::RBX => rbx,
            Register::RSP => rsp,
            Register::RBP => rbp,
            Register::RSI => rsi,
            Register::RDI => rdi,
            Register::R8 => r8,
            Register::R9 => r9,
            Register::R10 => r10,
            Register::R11 => r11,
            Register::R12 => r12,
            Register::R13 => r13,
            Register::R14 => r14,
            Register::R15 => r15,
            _ => panic!("Unsupported register"),
        }
    }
}

impl IAmd64Assembler for IcedAmd64Assembler {
    fn push_reg(&mut self, reg: Register) {
        self.assembler.push(Self::conv(reg)).unwrap();
    }

    fn push_mem64(&mut self, base_reg: Register, offset: i32) {
        self.assembler
            .push(qword_ptr(Self::conv(base_reg) + offset))
            .unwrap();
    }

    fn pop_reg(&mut self, reg: Register) {
        self.assembler.pop(Self::conv(reg)).unwrap();
    }

    fn mov_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.assembler
            .mov(Self::conv(reg1), Self::conv(reg2))
            .unwrap();
    }

    fn mov_reg_mem64(&mut self, dst_reg: Register, base_reg: Register, offset: i32) {
        self.assembler
            .mov(
                Self::conv(dst_reg),
                qword_ptr(Self::conv(base_reg) + offset),
            )
            .unwrap();
    }

    fn mov_mem64_reg(&mut self, base_reg: Register, offset: i32, src_reg: Register) {
        self.assembler
            .mov(
                qword_ptr(Self::conv(base_reg) + offset),
                Self::conv(src_reg),
            )
            .unwrap();
    }

    fn movabs_reg_imm64(&mut self, reg: Register, imm: u64) {
        self.assembler.mov(Self::conv(reg), imm).unwrap();
    }

    fn add_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.assembler
            .add(Self::conv(reg1), Self::conv(reg2))
            .unwrap();
    }

    fn add_reg_imm32(&mut self, reg: Register, imm32: u32) {
        self.assembler.add(Self::conv(reg), imm32 as i32).unwrap();
    }

    fn sub_reg_imm32(&mut self, reg: Register, imm32: u32) {
        self.assembler.sub(Self::conv(reg), imm32 as i32).unwrap();
    }

    fn imul_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.assembler
            .imul_2(Self::conv(reg1), Self::conv(reg2))
            .unwrap();
    }

    fn and_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.assembler
            .and(Self::conv(reg1), Self::conv(reg2))
            .unwrap();
    }

    fn and_reg_imm32(&mut self, reg: Register, imm: u32) {
        self.assembler.and(Self::conv(reg), imm as i32).unwrap();
    }

    fn and_mem64_reg(&mut self, base_reg: Register, offset: i32, src_reg: Register) {
        self.assembler
            .and(
                qword_ptr(Self::conv(base_reg) + offset),
                Self::conv(src_reg),
            )
            .unwrap();
    }

    fn or_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.assembler
            .or(Self::conv(reg1), Self::conv(reg2))
            .unwrap();
    }

    fn xor_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.assembler
            .xor(Self::conv(reg1), Self::conv(reg2))
            .unwrap();
    }

    fn not_reg(&mut self, reg: Register) {
        self.assembler.not(Self::conv(reg)).unwrap();
    }

    fn shl_reg_cl(&mut self, reg: Register) {
        self.assembler.shl(Self::conv(reg), cl).unwrap();
    }

    fn shr_reg_cl(&mut self, reg: Register) {
        self.assembler.shr(Self::conv(reg), cl).unwrap();
    }

    fn shr_reg_imm8(&mut self, reg: Register, imm8: u8) {
        self.assembler.shr(Self::conv(reg), imm8 as u32).unwrap();
    }

    fn call_reg(&mut self, reg: Register) {
        self.assembler.call(Self::conv(reg)).unwrap();
    }

    fn ret(&mut self) {
        self.assembler.ret().unwrap();
    }

    fn get_instructions(&mut self) -> Vec<Instruction> {
        self.assembler.instructions().to_vec()
    }

    fn get_bytes(&mut self) -> Vec<u8> {
        self.assembler.assemble(0).unwrap()
    }

    fn reset(&mut self) {
        self.assembler.reset();
    }
}

File: Mba.Simplifier/Jit/IcedAmd64Assembler.cs
using Iced.Intel;
using Mba.Common.MSiMBA;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static Iced.Intel.AssemblerRegisters;

namespace Mba.Simplifier.Jit
{
    public class IcedAmd64Assembler : IAmd64Assembler
    {
        private readonly Assembler assembler;

        public IcedAmd64Assembler(Assembler assembler)
        {
            this.assembler = assembler;
        }

        public List<byte> GetBytes()
        {
            return JitUtils.EncodeInstructions(GetInstructions(), 0, out ulong endRip).ToList();
        }

        public List<Instruction> GetInstructions()
        {
            return assembler.Instructions.ToList();
        }

        public void PushReg(Register reg)
        {
            assembler.push(Reg.Conv(reg));
        }

        public void PopReg(Register reg)
        {
            assembler.pop(Reg.Conv(reg));
        }

        public void MovRegReg(Register reg1, Register reg2)
        {
            assembler.mov(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void MovMem64Reg(Register baseReg, int offset, Register srcReg)
        {
            assembler.mov(__qword_ptr[Reg.Conv(baseReg) + offset], Reg.Conv(srcReg));
        }

        public void MovRegMem64(Register dstReg, Register baseReg, int offset)
        {
            assembler.mov(Reg.Conv(dstReg), __qword_ptr[Reg.Conv(baseReg) + offset]);
        }

        public void MovabsRegImm64(Register reg1, ulong imm)
        {
            assembler.mov(Reg.Conv(reg1), imm);
        }

        public void AddRegReg(Register reg1, Register reg2)
        {
            assembler.add(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void AddRegImm32(Register reg1, uint imm32)
        {
            assembler.add(Reg.Conv(reg1), (int)imm32);
        }

        public void SubRegImm32(Register reg1, uint imm32)
        {
            assembler.sub(Reg.Conv(reg1), (int)imm32);
        }

        public void ImulRegReg(Register reg1, Register reg2)
        {
            assembler.imul(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void AndRegReg(Register reg1, Register reg2)
        {
            assembler.and(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void AndRegImm32(Register reg1, uint imm32)
        {
            assembler.and(Reg.Conv(reg1), (int)imm32);
        }

        public void AndMem64Reg(Register baseReg, int offset, Register srcReg)
        {
            assembler.and(__qword_ptr[Reg.Conv(baseReg) + offset], Reg.Conv(srcReg));
        }

        public void OrRegReg(Register reg1, Register reg2)
        {
            assembler.or(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void XorRegReg(Register reg1, Register reg2)
        {
            assembler.xor(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void NotReg(Register reg1)
        {
            assembler.not(Reg.Conv(reg1));
        }

        public void ShlRegCl(Register reg1)
        {
            assembler.shl(Reg.Conv(reg1), cl);
        }

        public void ShrRegCl(Register reg1)
        {
            assembler.shr(Reg.Conv(reg1), cl);
        }

        public void ShrRegImm8(Register reg, byte imm8)
        {
            assembler.shr(Reg.Conv(reg), imm8);
        }

        public void CmpRegReg(Register reg1, Register reg2)
        {
            assembler.cmp(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void TestRegReg(Register reg1, Register reg2)
        {
            assembler.test(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void CmoveRegReg(Register reg1, Register reg2)
        {
            assembler.cmove(Reg.Conv(reg1), Reg.Conv(reg2));
        }

        public void SeteReg(Register reg1)
        {
            assembler.sete(Reg.Conv8(reg1));
        }

        public void SetneReg(Register reg1)
        {
            assembler.setne(Reg.Conv8(reg1));
        }

        public void SetaReg(Register reg1)
        {
            assembler.seta(Reg.Conv8(reg1));
        }

        public void SetbReg(Register reg1)
        {
            assembler.setb(Reg.Conv8(reg1));
        }

        public void CallReg(Register reg1)
        {
            assembler.call(Reg.Conv(reg1));
        }

        public void Ret()
        {
            assembler.ret();
        }

        public void Reset()
        {
            assembler.Reset();
        }

        public void PushMem64(Register baseReg, int offset)
        {
            assembler.push(__qword_ptr[Reg.Conv(baseReg) + offset]);
        }
    }
}
